app/controllers/application_controller.rb

```diff
@@ -1,10 +1,156 @@
 require './config/environment'
 
 class ApplicationController < Sinatra::Base
-
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    use Rack::Flash
+    set :session_secret, "ek_and_mk_for_lyfe"
+  end
+
+  helpers do
+    def current_user
+      User.find(session[:user_id])
+    end
+
+    def is_logged_in?(session)
+      session.has_key?(:user_id)
+    end
+  end
+
+  get '/' do
+    erb :index
+  end
+
+  get '/tweets' do
+    if is_logged_in?(session)
+      @tweets = Tweet.all.sort_by {|tweet| tweet.id}.reverse
+      erb :'/tweets/tweets'
+    else
+      flash[:error] = "Please log in to view tweets."
+      redirect '/login'
+    end
+  end
+
+  get '/tweets/new' do
+    if is_logged_in?(session)
+      erb :'/tweets/create_tweet'
+    else
+      flash[:error] = "Please log in or create an account to start tweeting."
+      redirect '/login'
+    end
+  end
+
+  post '/tweets' do
+    if params[:content] != ""
+      Tweet.create(content: params[:content], user_id: session[:user_id])
+      flash[:success] = "Tweet posted successfully!"
+      redirect '/tweets'
+    else
+      flash[:error] = "Tweets are short, but not that short. Make sure you have text in your tweet!"
+      redirect '/tweets/new'
+    end
+  end
+
+  get '/tweets/:id' do
+    if is_logged_in?(session)
+      @tweet = Tweet.find_by_id(params[:id])
+      erb :'/tweets/show_tweet'
+    else
+      flash[:error] = "Please log in to view tweets."
+      redirect '/login'
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    @tweet = Tweet.find_by_id(params[:id])
+    if is_logged_in?(session) && @tweet.user_id == current_user.id
+      erb :'/tweets/edit'
+    elsif is_logged_in?(session) && @tweet.user_id != current_user.id
+      flash[:error] = "Hands off that tweet! Remember, you can only edit your own tweets."
+      redirect '/tweets'
+    else
+      flash[:error] = "Please log in before editing any tweets."
+      redirect '/login'
+    end
+  end
+
+  patch '/tweets/:id' do
+    tweet = Tweet.find_by_id(params[:id])
+    if params[:content] != ""
+      tweet.update(content: params[:content])
+      flash[:success] = "Tweet updated successfully!"
+      redirect "/tweets/#{tweet.id}"
+    else
+      flash[:error] = "Tweets are short, but not that short. Make sure you have text in your tweet!"
+      redirect "/tweets/#{tweet.id}/edit"
+    end
+  end
+
+  delete '/tweets/:id/delete' do
+    tweet = Tweet.find_by_id(params[:id])
+    if is_logged_in?(session) && tweet.user_id == current_user.id
+      tweet.delete
```

Comment:
Another alternative is to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.